lib/membrane_flac_plugin/membrane_flac_parser.ex

```diff
@@ -65,45 +96,151 @@ defmodule Membrane.FLAC.Parser do
               {:stream_format, {:output, format}}
 
             %Buffer{} = buf ->
-              {:buffer, {:output, set_buffer_pts(buf, state)}}
+              {:buffer, {:output, buf}}
           end)
 
+        {actions, state} = calculate_pts(actions, state)
+
         {actions, %{state | parser: parser}}
 
       {:error, reason} ->
         raise "Parsing error: #{inspect(reason)}"
     end
   end
 
+  defp validate_pts_integrity(state)
+       when not state.generate_best_effort_timestamps? and
+              state.parser.queue != <<>> and
+              state.meta_queue == [] and
+              state.current_pts != nil and
+              state.input_pts != nil and
+              state.frame_duration != nil do
+    epsilon = state.frame_duration / 10
+
+    if state.input_pts < state.current_pts - epsilon do
+      Membrane.Logger.warning("PTS values are overlapping")
+    end
+
+    if state.input_pts > state.current_pts + epsilon do
+      Membrane.Logger.warning("PTS values are not continous")
+    end
+
+    :ok
+  end
+
+  defp validate_pts_integrity(_state), do: :ok
+
   @impl true
   def handle_end_of_stream(:input, _ctx, state) do
     {:ok, buffer} = Engine.flush(state.parser)
 
-    actions = [
-      buffer: {:output, set_buffer_pts(buffer, state)},
-      end_of_stream: :output,
-      notify_parent: {:end_of_stream, :input}
-    ]
+    {buffers_with_pts, state} = calculate_pts({:buffer, {:output, buffer}}, state)
+
+    actions =
+      buffers_with_pts ++
+        [
+          end_of_stream: :output,
+          notify_parent: {:end_of_stream, :input}
+        ]
 
     {actions, state}
   end
 
-  # set_buffer_pts() does not take into account changing sample rate during stream, because parser engine doesn't support it.
+  defp calculate_pts(actions, state) do
+    # in some cases actions are not a list but a single action tuple
+    actions =
+      if is_list(actions) do
+        actions
+      else
+        [actions]
+      end
+
+    # separate audio buffers from FLAC metadata buffers
+    {meta_buffers, audio_buffers} =
```

Comment:
Function passed to `Enum.split_with/2` below suggests, that one of these lists may contain some stream formats, but names of the variables suggest that both lists contain only buffers